[FLOC-1078] [FLOC-1473] [FLOC-1475] Add Ubuntu to installation instructions

jongiddy opened the pull request:
As well as adding Ubuntu instructions, uses a consistent prompt format throughout.

Still contains reference to SELinux, which will be modified by PR #1290

adamtheturtle commented on flocker/provision/_install.py:
```diff
@@ -128,15 +164,39 @@ def task_install_digitalocean_kernel():
+            # Update to read package info from new repos
+            run_from_args([
+                "apt-get", "update"]),
+            # Not clear that an upgrade is required at this point, so
```
Remove this unless an upgrade causes some acceptance tests to pass on a supported platform which don't pass otherwise.

adamtheturtle commented on flocker/provision/_install.py:
```diff
@@ -128,15 +164,39 @@ def task_install_digitalocean_kernel():
+def task_install_requirements(distribution):
+    if distribution == 'ubuntu-14.04':
+        return sequence([
+            # Add ZFS repo for recent ZFS versions - XXX no minimum
```
Link to an issue.

jongiddy replied:
Looking into it further, it seems Ubuntu does not have ZFS and Docker available without these commands - i.e. it is not just outdated versions, so I have changed the comments

adamtheturtle commented on flocker/provision/_install.py:
```diff
@@ -128,15 +164,39 @@ def task_install_digitalocean_kernel():
+            # version documented
+            run_from_args([
+                "add-apt-repository", "-y", "ppa:zfs-native/stable"]),
+            # Add Docker repo for recent Docker versions - XXX no
```
Link to an issue.

adamtheturtle commented on flocker/provision/_install.py:
```diff
@@ -128,15 +164,39 @@ def task_install_digitalocean_kernel():
+            # leave it out.
+            # XXX This brings up a prompt about upgrading grub,
+            # somehow work around that, see
+            # http://askubuntu.com/questions/187337/unattended-grub-configuration-after-kernel-upgrade
```
Am I right in thinking that `export DEBIAN_FRONTEND=noninteractive` works on some systems but not others?

adamtheturtle commented:
Unfortunately it seems that the issue was not clear enough. Not your fault, but this blocks https://clusterhq.atlassian.net/browse/FLOC-1100 "Install Flocker cluster on Ubuntu 14.04 for production".

The CLI work should be against (a blocker to) https://clusterhq.atlassian.net/browse/FLOC-1103, and should follow the acceptance criteria there.

See acceptance criteria on https://clusterhq.atlassian.net/browse/FLOC-1100. For example, after installation " a link to the configuration guide in the online documentation should be printed.".

However that also includes:

> Doc requirements: 
> * User must be guided to use these packages from the quickstart guide.

which seems more like FLOC-1103.

adamtheturtle commented on flocker/provision/_install.py:
```diff
@@ -72,15 +72,50 @@ def task_install_ssh_key():
+            run_from_args(['sync']),
+        ])
+    elif distribution == 'ubuntu-14.04':
+        # When 15.04 is available then the kernel can be backported from that,
```
This is due April 23rd. Follow up issue? Or if it'd be bad to ship this as-is maybe we should wait.

adamtheturtle commented:
The prompt work seems big enough that it should really be separated.

jongiddy commented:
@adamtheturtle So, the CLI stuff should be reverted and added as a separate issue blocking FLOC-1103?

adamtheturtle commented:
@jongiddy Yes

jongiddy commented:
At this point, separating out the work specifically required for FLOC-1100 and some of the formatting niceties would be a non-trivial task.

adamtheturtle commented on docs/indepth/installation.rst:
```diff
@@ -23,51 +23,53 @@ Linux
+On CentOS 7 or Fedora 20 you can install these by running:
 
-   alice@mercury:~$ sudo yum install @buildsys-build python python-devel python-virtualenv
+.. prompt:: bash linux$
```
The prompts should be consistent throughout the docs, as much as possible. `alice@mercury` is elsewhere and seems to be the standard in our docs for commands run on a development machine.

adamtheturtle commented:
I'll review the Ubuntu and prompt changes then. If the CLI work were as desired then I'd maybe say keep it, but because it uses `pip` instead of `deb`s it is worth separating.

adamtheturtle commented:
What I was going to say which may be useful in saving you time in follow-up work, was that the `VersionPrompt` runs into https://clusterhq.atlassian.net/browse/FLOC-1548 worse than ever, because the expected output does not show. This would be a regression in the docs not worth merging.

Firefox:

![screenshot 2015-04-08 15 58 55](https://cloud.githubusercontent.com/assets/797801/7055750/ee489284-de3d-11e4-9692-b6595fead21e.png)

Chrome:

![screenshot 2015-04-08 15 58 46](https://cloud.githubusercontent.com/assets/797801/7055755/f5864d5c-de3d-11e4-801c-a4306435c55f.png)

adamtheturtle commented on docs/indepth/installation.rst:
```diff
@@ -173,49 +175,49 @@ Using Amazon Web Services
 #. Add the *Key* to your local key chain (download it from the AWS web interface first if necessary):
 
-   .. code-block:: sh
+   .. prompt:: bash laptop$
```
This is better than it was, but annoyingly exposes us more to https://clusterhq.atlassian.net/browse/FLOC-1548 and we still don't have consistent prompts everywhere. I'll leave the judgement of which prompt to use, but I think that for now it should be the most common one. I think that might be `alice@mercury`. A follow-up can be made to improve them all.

adamtheturtle commented on docs/indepth/installation.rst:
```diff
@@ -225,7 +227,7 @@ Using Amazon Web Services
 
 
-#. Follow the :ref:`generic Fedora 20 installation instructions <fedora-20-install>` below.
+#. Follow the operating system specific installation instructions below.
```
This implies that we support all the operating systems below on DO, but we only support Fedora 20 for now, right?

adamtheturtle commented on flocker/provision/_install.py:
```diff
@@ -128,15 +163,39 @@ def task_install_digitalocean_kernel():
-            "yum", "install", "-y", "epel-release"]),
-        run_from_args(['sync']),
-    ])
+def task_install_requirements(distribution):
```
Docstring

adamtheturtle commented on docs/indepth/installation.rst:
```diff
@@ -173,49 +175,49 @@ Using Amazon Web Services
-   .. task:: upgrade_kernel
-      :prompt: [root@aws]#
+   .. task:: upgrade_kernel centos-7
+      :prompt: [root@centos]#
```
This being `[root@centos]` and the next one (same machine, same user) being `[root@aws]` is weird.

adamtheturtle commented on docs/indepth/installation.rst:
```diff
@@ -173,49 +175,49 @@ Using Amazon Web Services
+      ssh-add ~/.ssh/my-instance.pem
 
-#. Look up the public DNS name or public IP address of the new instance and log in as user "fedora", e.g.:
+#. Look up the public DNS name or public IP address of the new instance and, depending on the OS, log in as user "fedora", "centos", or "ubuntu" e.g.:
```
Same as the comment about DO - I think that this is useful work to keep for when we support non-Fedora on cloud providers, but that is not now.

adamtheturtle commented on docs/indepth/installation.rst:
```diff
@@ -359,19 +361,43 @@ The following commands will install the two repositories and the ``flocker-node`
+
+.. note:: The following commands all need to be run as root on the machine where ``clusterhq-flocker-node`` will be running.
+
+.. To test this:
```
:question: AFAIK you have been testing this with a different setup. If you think it'd be useful document that (in note form?).

adamtheturtle commented:
Thanks @jongiddy . I have left some feedback but the meat of the issue is something we did together and I'd like it reviewed by someone else, once you have addressed my feedback.

jongiddy commented:
When there aren't unrelated intermittent database problems in the tests, this is working for Ubuntu on Rackspace.

tomprince commented on docs/indepth/installation.rst:
```diff
@@ -327,21 +329,31 @@ The following commands will install the two repositories and the ``clusterhq-flo
 Installing on CentOS 7
 ----------------------
 
-Flocker requires the latest available kernel.
```
I'm curious how come you removed these instructions?

jongiddy replied:
Due to the different requirements upgrading on each cloud platform, it was easier to move the instructions for upgrading to the cloud-specific section.  Hence, once you've completed the cloud provisioning section, you should have the an appropriate kernel that has been rebooted if necessary.

The OS-specific section then starts from that point.

tomprince commented:
> uses a consistent prompt format throughout.

This makes it harder to review this branch. :frowning:

tomprince commented on flocker/provision/_install.py:
```diff
@@ -254,43 +302,83 @@ def task_install_flocker(
+                "add-apt-repository", "-y", "ppa:zfs-native/stable"]),
+            # Add Docker repo for recent Docker versions
+            run_from_args([
+                "add-apt-repository", "-y", "ppa:james-page/docker"]),
```
This will want to install the repository from #1287 as well.

jongiddy replied:
Added a TODO comment until repo is ready

tomprince commented on flocker/provision/_install.py:
```diff
@@ -254,43 +302,83 @@ def task_install_flocker(
+            run_from_args([
+                "apt-get", "update"]),
+            # Package spl-dkms sometimes does not have libc6-dev as a
+            # dependency, add it before ZFS installation requires it.
```
Link to a ticket.
Also, maybe this is worth including as `comment()` in the instructions.

jongiddy replied:
Linked to ticket

tomprince commented on flocker/provision/_install.py:
```diff
@@ -254,43 +302,83 @@ def task_install_flocker(
+            # Install Flocker node and all dependencies
+            run_from_args([
+                'apt-get', '-y', '--force-yes', 'install',
+                'clusterhq-flocker-node']),
```
This should support specifying the package version to install.

jongiddy replied:
Done

tomprince commented on flocker/provision/_install.py:
```diff
@@ -193,50 +204,87 @@ def configure_firewalld(rule):
+            for service in ['flocker-control-api', 'flocker-control-agent']
+        ])
+    elif distribution == 'ubuntu-14.04':
+        return sequence([])
```
I take it that ufw isn't enabled by default on rackspace or aws?

jongiddy replied:
It's true that UFW doesn't appear to be enabled anywhere by default.  However, I'm adding some changes to set UFW rules on Ubuntu.  As long as UFW is installed, which seems to be the case for all but most minimal systems (e.g. Docker), the firewall commands will succeed even if the firewall is not enabled.

tomprince commented:
The code changes look good modulo the above comments. There are enough unrelated documentation changes that I wasn't able to review them. Please resubmit after addressing the above points.

jongiddy commented:
Although there are doc changes on the installation page outside the Ubuntu section, they are pretty important. Dropping an Ubuntu section straight into a page originally developed for Fedora makes very little sense without restructuring to make it consistent throughout.

itamarst commented on admin/acceptance.py:
```diff
@@ -122,6 +125,14 @@ def stop_nodes(reactor):
 
 
+@implementer(INode)
+@attributes('address', 'distribution', apply_immutable=True)
```
Doesn't `@attributes` take a list of attribute names as an argument?

jongiddy replied:
Fixed

itamarst commented on admin/package-files/upstart/flocker-agent.conf:
```diff
@@ -0,0 +1,14 @@
+
+script
+	. /etc/default/flocker-agent.conf
+	exec /usr/sbin/flocker-zfs-agent $FLOCKER_NODE_NAME $FLOCKER_CONTROL_NODE
```
`FLOCKER_NODE_NAME` argument is no longer necessary, remove it.

jongiddy replied:
@itamarst It is still present in `master` in the systemd files.  Should it be removed from both Upstart and Systemd files?

jongiddy replied:
Looks like you are referring to FLOC-1631, which has not yet been merged.

jongiddy replied:
PR #1327 has now been merged, so this has been changed

itamarst commented on admin/packaging.py:
```diff
@@ -921,6 +932,9 @@ def omnibus_package_builder(
                     # SystemD configuration
                     package_files.child('systemd'):
                         FilePath('/usr/lib/systemd/system'),
+                    # Upstart configuration
```
:question: Do we want to install upstart stuff in RPM packages? I can see doing systemd in debs cause that's where Ubuntu is moving towards. Dropping in a few extra files isn't a big deal though.

jongiddy replied:
Yes, I think it is unusual, but it is the way it works right now. I have created issue FLOC-1736 to look at fixing it.

itamarst commented on flocker/provision/_install.py:
```diff
@@ -156,87 +155,151 @@ def task_install_kernel_devel():
+            run_from_args(['systemctl', 'enable', 'flocker-control']),
+            run_from_args(['systemctl', 'start', 'flocker-control']),
+        ])
+    elif distribution == 'ubuntu-14.04':
```
File an issue to make this less hacky.

jongiddy replied:
@itamarst Added FLOC-1735. Does this capture it OK?

itamarst commented:
I'd probably be fine merging this given some green builders, and just iterate on improvements since it's gotten rather sprawling. Unfortunately I can't review docs (buildbot links are unhappy), and it's very definitely not green... so I suggest:

1. Merge forward.
2. Add comments describing any follow up issues you want/need to do after this.
3. Make sure there's an Ubuntu acceptance tests builder that is green as a result of this branch.

At that point even if it's not perfect we can hopefully proceed with some confidence that at least it's a step in the right direction.

jongiddy commented:
Thanks for all the comments.

1. I have merged forward and updated Ubuntu to include the new `flocker-container-agent` service.
2. I added issue FLOC-1735 to tidy up the copious `if distribution...` lines.
3. The acceptance test we are using to evaluate this patch is `flocker/acceptance/rackspace/ubuntu-14.04`, which is usually green for this branch (barring total memory limit breaches on Rackspace). And `flocker-docs` is merely its usual orange for `build-linkcheck`.

adamtheturtle commented on docs/indepth/installation.rst:
```diff
@@ -350,29 +362,36 @@ The following commands will install the two repositories and the ``flocker-node`
 
-Installing ``flocker-node`` will automatically install Docker, but the ``docker`` service may not have been enabled or started.
-To enable and start Docker, run the following commands in a root console:
+.. To test this:
```
Remove this, or document it somewhere more appropriate.

jongiddy replied:
Removed, since acceptance test is now available

adamtheturtle commented on docs/spelling_wordlist.txt:
```diff
@@ -7,6 +7,7 @@ Btrfs
 bugfix
 Buildbot
 CentOS
+centos
```
Instead of these use double backticks around the usernames.

jongiddy replied:
Done.

adamtheturtle commented on flocker/provision/_install.py:
```diff
@@ -156,89 +155,154 @@ def task_install_kernel_devel():
+            run_from_args(['service', 'flocker-control', 'start']),
+        ])
+    else:
+        raise NotImplementedError()
```
No need for now - this doesn't exist elsewhere and will be ripped out for FLOC-1670.

tomprince replied:
This does in fact exist in numerous places already.

adamtheturtle commented on flocker/provision/_install.py:
```diff
@@ -256,43 +320,96 @@ def task_install_flocker(
+                "apt-get", "update"]),
+            # Package spl-dkms sometimes does not have libc6-dev as a
+            # dependency, add it before ZFS installation requires it.
+            # See https://github.com/zfsonlinux/zfs/issues/3298
```
This ticket is closed as a dupe of https://github.com/zfsonlinux/pkg-zfs/issues/129 which is itself closed and looks like it won't be re-opened.

tomprince replied:
We should still link to it, to explain why we are installing the package.

tomprince commented on flocker/provision/_rackspace.py:
```diff
@@ -32,7 +32,7 @@ def provision_rackspace(node, package_source, distribution, variants):
             address=node.address,
             commands=sequence([
-                task_upgrade_kernel_centos(),
+                task_upgrade_kernel('centos-7'),
```
We should pass in the distribution here, in case the `in (...)` changes.

tomprince commented:
There are a couple of test failures: http://build.clusterhq.com/builders/flocker-osx-10.10/builds/1602/steps/trial/logs/problems
